sycamore/transforms/extract_schema.py

```diff
@@ -0,0 +1,216 @@
+
+        if self._llm.is_chat_mode:
+            prompt = SCHEMA_ZERO_SHOT_GUIDANCE_PROMPT_CHAT
+
+        else:
+            prompt = SCHEMA_ZERO_SHOT_GUIDANCE_PROMPT
+
+        entities = self._llm.generate(
+            prompt_kwargs={"prompt": prompt, "entity": self._entity_name, "query": self._prompt_formatter(sub_elements)}
+        )
+
+        return entities
+
+
+class OpenAIPropertyExtractor(PropertyExtractor):
+    """
+    OpenAISchema uses one of OpenAI's language model (LLM) to extract property values once schema is established.
+    """
+
+    def __init__(
+        self,
+        # properties: list[str],
+        llm: LLM,
+        num_of_elements: int = 10,
+        prompt_formatter: Callable[[list[Element]], str] = element_list_formatter,
+    ):
+        super().__init__()
+        self._llm = llm
+        self._num_of_elements = num_of_elements
+        self._prompt_formatter = prompt_formatter
+
+    def extract_properties(self, document: Document) -> Document:
+        entities = self._handle_zero_shot_prompting(document)
+
+        try:
+            payload = entities["answer"]
+            answer = extract_json(payload)
+        except (json.JSONDecodeError, AttributeError):
+            answer = entities["answer"]
+
+        properties = document.properties
+        properties.update({"entity": answer})
+        document.properties = properties
+
+        return document
+
+    def _handle_zero_shot_prompting(self, document: Document) -> Any:
+        if document.text_representation:
+            text = document.text_representation
+        else:
+            text = self._prompt_formatter(
+                [document.elements[i] for i in range((min(self._num_of_elements, len(document.elements))))]
+            )
+
+        if self._llm.is_chat_mode:
+            prompt = PROPERTIES_ZERO_SHOT_GUIDANCE_PROMPT_CHAT
+        else:
+            prompt = PROPERTIES_ZERO_SHOT_GUIDANCE_PROMPT
+
+        schema_name = document.properties["_schema_class"]
+        schema = document.properties["_schema"]
+
+        entities = self._llm.generate(
+            prompt_kwargs={"prompt": prompt, "entity": schema_name, "properties": schema, "query": text}
+        )
+        return entities
+
+
+class ExtractSchema(Transform):
+    """
+    ExtractSchema is a transformation class for extracting a schema from a document using an SchemaExtractor.
+    """
+
+    def __init__(
+        self,
+        child: Node,
+        schema_extractor: SchemaExtractor,
+        **resource_args,
+    ):
+        super().__init__(child, **resource_args)
+        self._schema_extractor = schema_extractor
+
+    def execute(self) -> "Dataset":
+        input_dataset = self.child().execute()
+        dataset = input_dataset.map(generate_map_function(self._schema_extractor.extract_schema))
+        return dataset
+
+
+class ExtractBatchSchema(Transform):
```

Comment:
I'm having a hard time really grokking the batch vs non-batch implementation. Maybe you can walk me through it when we talk.